central/auth/m2m/role_mapper.go

```diff
@@ -0,0 +1,104 @@
+package m2m
+
+import (
+	"context"
+	"regexp"
+
+	"github.com/pkg/errors"
+	roleDataStore "github.com/stackrox/rox/central/role/datastore"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/auth/permissions"
+	"github.com/stackrox/rox/pkg/grpc/authn"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/set"
+)
+
+var (
+	_ permissions.RoleMapper = (*roleMapper)(nil)
+
+	log = logging.LoggerForModule()
+)
+
+type roleMapper struct {
+	config *storage.AuthMachineToMachineConfig
+	roleDS roleDataStore.DataStore
+}
+
+func newRoleMapper(config *storage.AuthMachineToMachineConfig, roleDS roleDataStore.DataStore) permissions.RoleMapper {
+	return &roleMapper{
+		config: config,
+		roleDS: roleDS,
+	}
+}
+
+func (r *roleMapper) FromUserDescriptor(ctx context.Context, user *permissions.UserDescriptor) ([]permissions.ResolvedRole, error) {
+	return resolveRolesForClaims(ctx, user.Attributes, r.roleDS, r.config.GetMappings())
+}
+
+func resolveRolesForClaims(ctx context.Context, claims map[string][]string, roleDS roleDataStore.DataStore,
+	mappings []*storage.AuthMachineToMachineConfig_Mapping) ([]permissions.ResolvedRole, error) {
+	rolesForUser := set.NewStringSet()
+
+	for _, mapping := range mappings {
+		if valuesMatch(mapping.GetValue(), claims[mapping.GetKey()]) {
+			rolesForUser.Add(mapping.GetRole())
+		}
+	}
```

Comment:
This makes it impossible to filter by GitHub repository name **and** some other claim. If any one matches, it becomes the role.

This comment is probably not actionable, it relates to my other comment about checking clientIDs.

Reply:
Discussed this offline, for posterity:

For the time being in 4.3 we agreed to keep the disjunction logic of the mappings.
In case feedback suggests there should be some change, it will be addressed in 4.4.

Reply:
Also part of the offline agreement is to _ask_ for this specific part of feedback. Like "how do you guys find ability to configure role mapping, is it sufficient or do you need to make more powerful expressions, like `AND` multiple claims?"

Feel free to resolve this thread.